File: app/api/personalize.py
"""Personalize / explainability layer.

Mirrors the `flagRules` block in `src/lib/api.ts#resolveNutrition`
exactly so the NutrientsScreen verdict chips render identical
contributions when wired to the backend. The math is the same;
only the language is Python — and now, the rule table is data.
"""

from __future__ import annotations

from dataclasses import dataclass

from app.models.schemas import (
    DetectedItem,
    HealthGoals,
    NutrientContribution,
    ResolvedItem,
    UserProfile,
)
from app.seed.fixtures import TARGETS


# A flagged nutrient is one the user's profile cares about.
# Diabetic profiles care about sugar + fiber; everyone else sodium + fiber.
_PROFILE_FOCUS: dict[bool, frozenset[str]] = {
    True: frozenset({"sugar", "fiber"}),
    False: frozenset({"sodium", "fiber"}),
}


@dataclass(frozen=True)
class _Rule:
    nutrient: str
    threshold: float
    low_threshold: float | None  # below this → "good" tone with low-reason
    warn_share: float  # meal-over + share ≥ this → "warn" tone with high-reason
    unit: str  # "g" or "mg" — formats the interpolation
    high_reason: str  # template; rendered with {name}, {amount}, {pct}
    low_reason: str | None  # template; None means "no low-reason comment"


_RULES: list[_Rule] = [
    _Rule(
        nutrient="sodium",
        threshold=TARGETS.sodium,
        low_threshold=None,
        warn_share=0.30,
        unit="mg",
        high_reason="{name} adds {amount}mg sodium — that's {pct}% of your daily cap in one item.",
        low_reason=None,
    ),
    _Rule(
        nutrient="sugar",
        threshold=TARGETS.sugar,
        low_threshold=TARGETS.sugar * 0.6,
        warn_share=0.30,
        unit="g",
        high_reason="{name} brings {amount}g of sugar — a notable spike for a prediabetic-friendly plate.",
        low_reason="{name} contributes only ~{amount}g of sugar — well within your prediabetic-friendly target.",
    ),
    _Rule(
        nutrient="fiber",
        threshold=TARGETS.fiber,
        low_threshold=12.0,
        warn_share=0.30,
        unit="g",
        high_reason="{name} is a solid fiber source at {amount}g — pulls the meal toward target.",
        low_reason="{name} contributes only ~{amount}g fiber — the whole meal falls short of your 30g daily target. Add whole grains or legumes next time.",
    ),
    _Rule(
        nutrient="protein",
        threshold=TARGETS.protein,
        low_threshold=None,
        warn_share=0.30,
        unit="g",
        high_reason="{name} delivers {amount}g of protein — strong for a single dish.",
        low_reason="{name} is light on protein (~{amount}g) — consider adding dal or paneer.",
    ),
]


def compute_contributions(
    resolved: list[ResolvedItem],
    profile: UserProfile | None,
) -> list[NutrientContribution]:
    """Per-item, per-nutrient contributions for the explainability panel.

    A nutrient is "in play" iff the profile cares about it. Within those,
    we flag two cases:
      • Meal exceeds the nutrient's high threshold AND one item owns
        ≥30% of the meal's share → "warn" tone with the high-reason.
      • Meal falls below the low threshold → "good" tone with the
        low-reason for each non-trivial item (share ≥ 10%).
    """
    if profile is None:
        return []
    focus = _PROFILE_FOCUS[profile.goals.diabetic]

    # Per-nutrient meal totals — exclude items whose nutrition we
    # couldn't resolve (they'd just add noise to the share calc).
    totals: dict[str, float] = {
        r.nutrient: sum(_amount(it, r.nutrient) for it in resolved)
        for r in _RULES
        if r.nutrient in focus
    }

    contributions: list[NutrientContribution] = []
    for rule in _RULES:
        if rule.nutrient not in focus:
            continue
        total = totals[rule.nutrient]
        over = total > rule.threshold
        under = rule.low_threshold is not None and total < rule.low_threshold
        if not (over or under):
            continue
        for item in resolved:
            if item.nutrition is None:
                continue
            amount = _amount(item, rule.nutrient)
            share = amount / total if total > 0 else 0.0
            if share < 0.10:
                continue
            if over and share >= rule.warn_share:
                contributions.append(_build(
                    rule, item, amount, share, flagged=True, tone="warn",
                ))
            elif not over and rule.low_reason is not None:
                contributions.append(_build(
                    rule, item, amount, share, flagged=False, tone="good",
                ))
    return contributions


def _amount(item: ResolvedItem, nutrient: str) -> float:
    """Grams-scaled contribution of one resolved item to one nutrient."""
    return getattr(item.nutrition, nutrient) * (item.grams / 100.0)


def _build(
    rule: _Rule,
    item: ResolvedItem,
    amount: float,
    share: float,
    *,
    flagged: bool,
    tone: str,
) -> NutrientContribution:
    template = rule.high_reason if tone == "warn" else rule.low_reason
    amount_str = (
        str(round(amount)) if rule.unit == "mg" else str(round(amount, 1))
    )
    kwargs: dict[str, object] = {"name": item.name, "amount": amount_str}
    if tone == "warn":
        kwargs["pct"] = (
            round((amount / rule.threshold) * 100) if rule.threshold > 0 else 0
        )
    return NutrientContribution(
        nutrient=rule.nutrient,  # type: ignore[arg-type]
        item_name=item.name,
        amount=amount,
        share=share,
        flagged=flagged,
        reason=template.format(**kwargs),  # type: ignore[misc]
        tone=tone,  # type: ignore[arg-type]
    )


def resolve_detected_item(item: DetectedItem, off_facts) -> ResolvedItem:
    """Convert a `DetectedItem` into a `ResolvedItem` by joining against
    the seeded `NUTRITION_DB` first, then falling back to Open Food Facts.

    `off_facts` is a callable `(name: str) -> NutritionFacts | None` —
    injected so this layer doesn't depend on the OFF client directly
    (testable in isolation).
    """
    # Imported here to avoid a circular import at module load.
    from app.seed.fixtures import NUTRITION_DB

    if (descriptor := NUTRITION_DB.get(item.name)) is not None:
        return ResolvedItem(
            **item.model_dump(), nutrition=descriptor, descriptor=descriptor,
            source="cache", partial=False,
        )
    if (facts := off_facts(item.name)) is not None:
        return ResolvedItem(
            **item.model_dump(), nutrition=facts, descriptor=None,
            source="open-food-facts", partial=False,
        )
    return ResolvedItem(
        **item.model_dump(), nutrition=None, descriptor=None,
        source="estimated", partial=True,
    )

File: src/lib/personalize.ts
/**
 * Personalize / explainability layer — single source of truth.
 *
 * The flag-rules table and the per-profile nutrient focus map live here
 * so the mock `api.ts#resolveNutrition` (frontend) and the NutrientsScreen
 * verdict UI can't drift. The backend mirrors the same shape in
 * `app/api/personalize.py`.
 */

import type { NutritionFacts, NutrientContribution, UserProfile } from '../types/schemas';
import { MOCK_TARGETS } from '../mocks/fixtures';

// ── Nutrient focus per profile ──────────────────────────────────────────
// Diabetic profiles care about sugar + fiber; everyone else sodium + fiber.
function focusFor(diabetic: boolean): ReadonlySet<string> {
  return diabetic ? new Set(['sugar', 'fiber']) : new Set(['sodium', 'fiber']);
}

export function isNutrientFocused(nutrient: string, profile: UserProfile | null | undefined): boolean {
  if (!profile) return false;
  return focusFor(profile.goals.diabetic).has(nutrient);
}

/** The list of flagged nutrients to render on NutrientsScreen (the "Why these flags?" strip). */
export function focusedNutrients(profile: UserProfile | null | undefined): ReadonlyArray<{ nutrient: 'sodium' | 'sugar' | 'fiber'; threshold: number }> {
  if (!profile) return [];
  return profile.goals.diabetic
    ? [
        { nutrient: 'sugar', threshold: MOCK_TARGETS.sugar },
        { nutrient: 'fiber', threshold: MOCK_TARGETS.fiber },
      ]
    : [
        { nutrient: 'sodium', threshold: MOCK_TARGETS.sodium },
        { nutrient: 'fiber', threshold: MOCK_TARGETS.fiber },
      ];
}

// ── Flag rules (the SHAP-style contribution engine) ─────────────────────

export type FlagReason = (name: string, amount: number) => string;

interface FlagRule {
  nutrient: NutrientContribution['nutrient'];
  threshold: number;
  reasonIfHigh: FlagReason;
  reasonIfLow?: FlagReason;
  lowThreshold?: number;
  /** "g" or "mg" — controls the formatting precision in the verdict. */
  unit: 'g' | 'mg';
}

export const FLAG_RULES: readonly FlagRule[] = [
  {
    nutrient: 'sodium',
    threshold: MOCK_TARGETS.sodium,
    unit: 'mg',
    reasonIfHigh: (n, a) =>
      `${n} adds ${Math.round(a)}mg sodium — that's ${Math.round((a / MOCK_TARGETS.sodium) * 100)}% of your daily cap in one item.`,
  },
  {
    nutrient: 'sugar',
    threshold: MOCK_TARGETS.sugar,
    unit: 'g',
    reasonIfHigh: (n, a) =>
      `${n} brings ${a.toFixed(1)}g of sugar — a notable spike for a prediabetic-friendly plate.`,
    reasonIfLow: (n, a) =>
      `${n} contributes only ~${a.toFixed(1)}g of sugar — well within your prediabetic-friendly target.`,
    lowThreshold: MOCK_TARGETS.sugar * 0.6,
  },
  {
    nutrient: 'fiber',
    threshold: MOCK_TARGETS.fiber,
    unit: 'g',
    reasonIfHigh: (n, a) =>
      `${n} is a solid fiber source at ${a.toFixed(1)}g — pulls the meal toward target.`,
    reasonIfLow: (n, a) =>
      `${n} contributes only ~${a.toFixed(1)}g fiber — the whole meal falls short of your 30g daily target. Add whole grains or legumes next time.`,
    lowThreshold: 12,
  },
  {
    nutrient: 'protein',
    threshold: MOCK_TARGETS.protein,
    unit: 'g',
    reasonIfHigh: (n, a) =>
      `${n} delivers ${a.toFixed(1)}g of protein — strong for a single dish.`,
    reasonIfLow: (n, a) =>
      `${n} is light on protein (~${a.toFixed(1)}g) — consider adding dal or paneer.`,
  },
];

/** The 7 nutrients the resolver sums over (per-100g scaling). */
export const NUTRIENT_KEYS = [
  'sodium', 'sugar', 'fiber', 'protein', 'carbs', 'fat', 'kcal',
] as const;

/** Sum one resolved item's grams-scaled contribution to a single nutrient. */
export function itemAmount(
  nutrition: NutritionFacts,
  nutrient: string,
  grams: number,
): number {
  return (nutrition as unknown as Record<string, number>)[nutrient] * (grams / 100);
}

/** Display unit for a given nutrient, read from the rule table. Mirrors the
 *  backend's per-nutrient unit conventions (sodium is milligrams, everything
 *  else is grams). Callers — e.g. NutrientsScreen — should prefer this over a
 *  hard-coded ternary so adding a new unit (mg for sugar, etc.) doesn't drift. */
export function unitForNutrient(nutrient: NutrientContribution['nutrient']): 'g' | 'mg' {
  const rule = FLAG_RULES.find((r) => r.nutrient === nutrient);
  return rule?.unit ?? 'g';
}

/**
 * Compute per-nutrient meal totals across all resolved items.
 * Items without `nutrition` are skipped — they can't contribute to the share.
 */
export function mealTotals<T extends { nutrition: NutritionFacts | null; grams: number }>(
  resolved: readonly T[],
): Record<typeof NUTRIENT_KEYS[number], number> {
  const totals = Object.fromEntries(NUTRIENT_KEYS.map((k) => [k, 0])) as Record<typeof NUTRIENT_KEYS[number], number>;
  for (const r of resolved) {
    if (!r.nutrition) continue;
    const factor = r.grams / 100;
    for (const k of NUTRIENT_KEYS) {
      totals[k] += (r.nutrition as unknown as Record<string, number>)[k] * factor;
    }
  }
  return totals;
}

/** Build a per-nutrient list of contributions, applying the rule table + profile focus. */
export function buildContributions(
  resolved: ReadonlyArray<{ name: string; grams: number; nutrition: NutritionFacts | null }>,
  totals: Record<string, number>,
  profile: UserProfile | null | undefined,
): NutrientContribution[] {
  const out: NutrientContribution[] = [];
  for (const rule of FLAG_RULES) {
    if (!isNutrientFocused(rule.nutrient, profile)) continue;
    const total = totals[rule.nutrient] ?? 0;
    const mealOver = total > rule.threshold;
    const mealUnder = rule.lowThreshold != null && total < rule.lowThreshold;
    if (!mealOver && !mealUnder) continue;
    for (const r of resolved) {
      if (!r.nutrition) continue;
      const amount = itemAmount(r.nutrition, rule.nutrient, r.grams);
      const share = total > 0 ? amount / total : 0;
      if (share < 0.10) continue;
      const tone: 'warn' | 'good' = mealOver && share >= 0.30 ? 'warn' : 'good';
      const reason = tone === 'warn'
        ? rule.reasonIfHigh(r.name, amount)
        : rule.reasonIfLow?.(r.name, amount);
      if (!reason) continue;
      out.push({
        nutrient: rule.nutrient,
        itemName: r.name,
        amount,
        share,
        flagged: tone === 'warn',
        reason,
        tone,
      });
    }
  }
  return out;
}